### DISCUSSION TITLE
Best way to have onCreate/onDestroy on a Node

### DISCUSSION THREAD

**Discussion by dimsuz** [bumble-tech/appyx#285] [Q&A] [answered]
What is the best way to have an `onCreate`/`onDestroy` handling inside a particular `Node`?

For example I have a `ParentNode` and I want to create a `CoroutineScope` which is created when it is inserted in the graph and is `cancel()`-ed when `Node` is removed from the graph.

And also this would have to be repeated in case I navigate to this node again later: new scope would have to be created and then destroyed on exit.

I want to be able to do this from _within_ the node itself, not from its parent.

**Comment by KovalevAndrey** [comment, marked as answer]
@dimsuz you can use `lifecycleScope` in the Node. It's created when the `Node` is created and the scope is cancelled automatically `onDestroy`

**Reply by dimsuz** [reply]
I see! 

But is there a generic mechanism for this? `CoroutineScope` was only one example I thought of. I also can have some different resource which I would like to clean up...

**Reply by KovalevAndrey** [reply]
@dimsuz if you need to clean up your resources, take look at `Destroyable` plugin. Let your classes implement this plugin and perform cleanup on `destroy()` method.  `Destroyable` interface must be passed to the list of `Node` plugins in the constructor

**Reply by dimsuz** [reply]
Nice, this should work, thank you!

I guess node constructor will act as `onCreate`, and destroyable plugin will act as `onDestroy`. 

Or perhaps I can have both of these "points" stuffed into a single Plugin implementation, I'll check. Will be nicer to have them in one place for clarity.

**Reply by KovalevAndrey** [reply]
@dimsuz It depends on your needs. `Node` is moved in `onCreate` state when `onBuilt()` is called. Sometimes doing thing in constructor is too early, and you can use `NodeLifecycleAware` plugin instead. `fun onCreate(lifecycle: Lifecycle)` is called when `Node` `onCreate` gets created. You can observe `lifecycle` events do cleanup `onDestroy` and more
